dspace-api/src/main/java/org/dspace/content/CollectionServiceImpl.java

```diff
@@ -757,6 +757,22 @@ public void delete(Context context, Collection collection) throws SQLException,
             }
         }
 
+        // Remove all workflow groups
+        Group workflowGroup = collection.getWorkflowStep1(context);
+		if (workflowGroup != null) {
+            collection.setWorkflowGroup(context, 1, null);
+            this.groupService.delete(context, workflowGroup);
```

Comment:
It appears that this is causing a large number of our Unit Tests to fail, starting with the `EPersonTest` method called ` testCascadingDeleteSubmitterPreservesWorkflowItems()`.

It looks like this group deletion results in a database error:
```
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: Referential integrity constraint violation: "CONSTRAINT_C3: PUBLIC.CWF_POOLTASK FOREIGN KEY(GROUP_ID) REFERENCES PUBLIC.EPERSONGROUP(UUID) (UUID '860be8f6-703d-4ec2-b771-51e9a1b8ecd8')"; SQL statement:
delete from PUBLIC.epersongroup where uuid=? [23503-210]
```

That is what is logged by our automated Unit Tests.  It likely can be reproduced locally by running the same Unit Tests locally, see: https://github.com/DSpace/DSpace/#running-tests

So, it appears this is not fully working.  Based on the error, it appears that a Workflow Pool Task still exists that references these groups.  Likely, if a Collection is deleted, all WorkflowItems and WorkspaceItems linked that that Collection may also need to be removed... or the Collection should not be possible to delete until someone manually deletes all WorkflowItems/WorkspaceItems using the "Administer Workflow" User Interface.

This issue seems to be getting more complex than I thought it would initially be.  So, it may require some more careful thinking & additional automated tests.

Reply:
As an aid to tracking down these dependencies, I tried exporting the database schema (`pg_dump --schema-only dspace > schema`) and searching for constraints (`grep REFERENCES.*epersongroup schema`).  The workflow system does touch a lot of (JPA) entity types.

We need to be careful not to delete groups that are used *outside of* the workflow system, too.  Nothing prevents an admin. granting rights in some arbitrary object to a workflow group (although it would be saner to make the workflow group a member of some other group and grant rights to *that* group).  Possibly the best thing to do would be to identify all of the references that can safely be revoked, try to delete the group, simply log any failure and continue.

I would suggest that workspace items can simply be unlinked from the group.  Workflow items can be returned to the submitter's workspace.  If an uninstalled item *is* deleted as part of deleting a collection, the submitter should receive an email explaining what happened to the submission.  (I am uncertain whether it is possible for the submitter to designate a new group for a returned submission.  Maybe "delete and email" would be best.)

I am concerned that I am seeing `getWorkflowStep1` etc in new code.  I don't think that plays well with the configurable workflow system.  We ought to deprecate those and schedule their removal, no?

Reply:
I am concerned that I am seeing `getWorkflowStep1` etc. in new code.  IIRC this doesn't necessarily work with the configurable workflow implementation.  We should schedule their removal.

As an aid to finding these dependencies, I tried exporting the database schema (`pg_dump --schema-only dspace > schema`) and searching for constraints (`grep REFERENCES.*epersongroup schema`).  Share and enjoy.  There may be other database entity types implicated as well.

Besides pool tasks, there may be claimed tasks for submissions that have entered workflow and accepted by editors.

I don't know whether there is a sensible way to let a submitter re-submit to a new group when the original group is gone.  If not, then uninstalled items which reference the group must be deleted.  I would suggest that the user receive an email notice of the item's deletion.

We should be careful not to remove a group that is also referenced outside of workflow.  Nothing prevents an administrator granting rights in some arbitrary item to a workflow group (although it would be saner to make the workflow group a member of another group and grant rights to *that* group).